shared/lib_battery_lifetime_nmc.cpp

```diff
@@ -0,0 +1,216 @@
+        lifetime_t(rhs){
+    operator=(rhs);
+}
+
+lifetime_nmc_t& lifetime_nmc_t::operator=(const lifetime_nmc_t& rhs) {
+    if (this != &rhs) {
+        *params = *rhs.params;
+        *state = *rhs.state;
+    }
+    return *this;
+}
+
+lifetime_t * lifetime_nmc_t::clone() {
+    return new lifetime_nmc_t(*this);
+}
+
+double lifetime_nmc_t::calculate_Uneg(double SOC) {
+    double Uneg;
+    if (SOC <= 0.1)
+        Uneg = ((0.2420 - 1.2868) / 0.1) * SOC + 1.2868;
+    else
+        Uneg = ((0.0859 - 0.2420) / 0.9) * (SOC - 0.1) + 0.2420;
+    return Uneg;
+}
+
+double lifetime_nmc_t::calculate_Voc(double SOC) {
+    double Voc;
+    if (SOC <= 0.1)
+        Voc = ((0.4679) / 0.1) * SOC + 3;
+    else if (SOC <= 0.6)
+        Voc = ((3.747 - 3.4679) / 0.5) * (SOC - 0.1) + 3.4679;
+    else
+        Voc = ((4.1934 - 3.7469) / 0.4) * (SOC - 0.6) + 3.7469;
+    return Voc;
+}
+
+double lifetime_nmc_t::runQli() {
+    double dt_day = 1;
+    int dn_cycles = state->n_cycles - state->nmc->n_cycles_prev_day;
+    double k_cal = 0;
+    //double b1 = std::accumulate(state->nmc->b1_dt.begin(), state->nmc->b1_dt.end(), 0);
+    double b1 = state->nmc->b1_dt;
+    double b2 = state->nmc->b2_dt;
+    double b3 = state->nmc->b3_dt;
+
+    state->nmc->b1_dt = 0;
+    state->nmc->b2_dt = 0;
+    state->nmc->b3_dt = 0;
+
+    if (state->day_age_of_battery > 0)
+        k_cal = (0.5 * b1) / (sqrt(state->day_age_of_battery)) + (b3 / tau_b3) * exp(-(state->day_age_of_battery / tau_b3));
+    else
+        k_cal = 0;
+
+    double dq_new;
+    if (state->nmc->dq_relative_li_old == 0)
+        dq_new = k_cal * dt_day + b2 * dn_cycles;
+    else
+        dq_new = k_cal * dt_day + b2 * dn_cycles + state->nmc->dq_relative_li_old;
+    state->nmc->dq_relative_li_old = dq_new;
+    state->nmc->q_relative_li = (1.07 - (dq_new)) * 100;
+    return state->nmc->q_relative_li;
+}
+
+double lifetime_nmc_t::runQneg(double T_battery, double SOC) {
+
+    int dn_cycles = state->n_cycles - state->nmc->n_cycles_prev_day;
+
+    double c2 = state->nmc->c2_dt;
+    state->nmc->c2_dt = 0;
+
+    double dq_new;
+    if (state->nmc->dq_relative_neg_old == 0)
+        dq_new = 1 - sqrt(1 - 2 * (c2 / c0_ref) * dn_cycles);
+    else
+        dq_new = 1 - sqrt(1 - 2 * (c2 / c0_ref) * dn_cycles) + state->nmc->dq_relative_neg_old;
+
+    state->nmc->dq_relative_neg_old = dq_new;
+
+    state->nmc->q_relative_neg = (1 - (dq_new)) * 100;
+
+    //return state->nmc->q_relative_neg;
+    return 100;
+}
+
+void lifetime_nmc_t::runLifetimeModels(size_t lifetimeIndex, bool charge_changed, double prev_DOD, double DOD,
+                                       double T_battery) {
+    double q_last = state->q_relative;
+
+    // update day age of battery
+    size_t day_age_of_battery_old = (size_t)(state->day_age_of_battery);
+    state->day_age_of_battery += params->dt_hr / (double)util::hours_per_day;
+    auto ts_per_day = (size_t)(util::hours_per_day / params->dt_hr);
```

Comment:
Using the lifetime index to calculate when a full day has passed, this is a bit more rigorous vs converting between a floating point and an int or size_t